sdk/python/toolchain/uv.go

```diff
@@ -0,0 +1,265 @@
+			if err != nil {
+				if !errors.Is(err, os.ErrNotExist) {
+					return nil, fmt.Errorf("error while looking for pyproject.toml in %s: %w", root, err)
+				}
+				// We have no uv.lock and no pyproject.toml, place the virtualenv in the project root.
+				virtualenv = filepath.Join(root, defaultVirtualEnv)
+			} else {
+				// We have a pyproject.toml, place the virtualenv next to it.
+				virtualenv = filepath.Join(pyprojectTomlDir, defaultVirtualEnv)
+			}
+		} else {
+			// We have a uv.lock, place the virtualenv next to it.
+			virtualenv = filepath.Join(uvLockDir, defaultVirtualEnv)
+		}
+	}
+	if !filepath.IsAbs(virtualenv) {
+		virtualenv = filepath.Join(root, virtualenv)
+	}
+
+	u := &uv{
+		virtualenvPath: virtualenv,
+		root:           root,
+	}
+
+	// Validate the version
+	cmd := u.uvCommand(context.Background(), "", false, nil, nil, "--version")
+	versionString, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get uv version: %w", err)
+	}
+	version, err := u.uvVersion(string(versionString))
+	if err != nil {
+		return nil, err
+	}
+	logging.V(9).Infof("Python toolchain: using uv version %s", version)
+
+	return u, nil
+}
+
+func (u *uv) InstallDependencies(ctx context.Context, cwd string, useLanguageVersionTools,
+	showOutput bool, infoWriter, errorWriter io.Writer,
+) error {
+	if useLanguageVersionTools {
+		if err := installPython(ctx, cwd, showOutput, infoWriter, errorWriter); err != nil {
+			return err
+		}
+	}
+
+	// Look for a uv.lock file.
+	// If no uv.lock file is found, look for a pyproject.toml file.
+	// If no pyproject.toml file is found, create it, and then look for a
+	// requirements.txt file to install dependencies.
+	if _, err := searchup(cwd, "uv.lock"); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("error while looking for uv.lock in %s: %w", cwd, err)
+		}
+
+		// No uv.lock found, look for pyproject.toml.
+		if _, err := searchup(cwd, "pyproject.toml"); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("error while looking for pyproject.toml in %s: %w", cwd, err)
+			}
+
+			// No pyproject.toml found, create one using `uv`
+			initCmd := u.uvCommand(ctx, cwd, showOutput, infoWriter, errorWriter,
+				"init", "--no-readme", "--no-package", "--no-pin-python")
+			if err := initCmd.Run(); err != nil {
+				return fmt.Errorf("error initializing python project: %w", err)
+			}
+
+			// `uv init` creates a `hello.py` file, delete it.
+			contract.IgnoreError(os.Remove(filepath.Join(cwd, "hello.py")))
+
+			// Look for a requirements.txt file and use it to install dependencies
+			requirementsTxtDir, err := searchup(cwd, "requirements.txt")
+			if err == nil {
+				requirementsTxt := filepath.Join(requirementsTxtDir, "requirements.txt")
+				addCmd := u.uvCommand(ctx, cwd, showOutput, infoWriter, errorWriter, "add", "-r", requirementsTxt)
+				if err := addCmd.Run(); err != nil {
```

Comment:
We're not doing this in most places currently, but I'm wondering if we should check if this error is an `ExitError` (which it usually will be), and and include [exitError.Stderr](https://pkg.go.dev/os/exec#ExitError) in the error we're returning?  Unfortunately the `Error` method of `ExitError` only includes the exit code afaiu, which is kind of annoying for debugging purposes.

Reply:
That's a really good idea. I added a `errorWithStderr` helper and used that for all error returns from `cmd.Run`. For now it's a private helper in the toolchain package, but I might follow up with moving that to a shared location and use it throughout the codebase.

Reply:
Awesome, yeah I think this would be great to be used throughout!